File: matlab/tools/generic/common_mode_noise_canceller.m
function [x_den, x_den_smoothed, s, alpha, bl] = common_mode_noise_canceller(x, fs, varargin)
%COMMON_MODE_NOISE_CANCELLER Estimates and removes common-mode noise from multichannel data
%
%   [x_den, x_den_smoothed, s, alpha] = common_mode_noise_canceller(x, fs, varargin)
%   estimates and removes common-mode noise from a given multichannel signal
%   matrix `x`, sampled at frequency `fs`. This function implements an iterative
%   algorithm to minimize the squared error between the observed signals and
%   their estimates, produced by common-mode noise removal. The assumption
%   is that a common noise of unknown amplitude has impacted all channels
%   of x.
%
%   Inputs:
%       x - A matrix representing multichannel data, with size N channels x T samples.
%       fs - Sampling frequency of the signal.
%       fc (optional) - Cutoff frequency for low-pass filtering. If not
%              specified, no filtering is applied. If set to 0, the mean
%              of `x` is subtracted per channel.
%       itr (optional) - Number of iterations for the algorithm. Default is 10.
%       lambda (optional) - Regularization parameter for a 2nd order Tikhonov
%              regularization. Default is 100. If set to 0, no
%              regularization is applied.
%       optim_indexes (optional) - Indices of the samples to optimize.
%              Default is all samples.
%       plot_results (optional) - Boolean to enable/disable result plotting.
%              Default is false (no plotting).
%
%   Outputs:
%       x_den - The denoised signal, with common-mode noise removed (N x T).
%       x_den_smoothed - The denoised signal after smoothing with Tikhonov regularization (N x T).
%       s - The estimated common-mode signal (1 x T).
%       alpha - The scaling factor or matrix used in noise cancellation (N x 1).
%       bl - channel-wise average or baseline estimated by high-pass
%           filtering (N x T) or empty, if no `fc` is set 
%
%   Method:
%   This function addresses common-mode noise cancellation in multichannel
%   data by using iterative least squares optimization on a data matrix `X`
%   (N channels by T samples). It aims to minimize the norm of the
%   difference between `X` and its estimate `X_hat` = alpha * s, where
%   `alpha` is an Nx1 vector and `s` is a 1xT vector representing the
%   common-mode signal. The optimization problem is formulated as
%   minimizing the Frobenius norm ||X - alpha * s||^2. The algorithm
%   initializes `s` as the median of `X` across channels and iteratively
%   updates `alpha` and `s` by applying least squares. It allows focusing
%   on specific segments of `X` through `optim_indexes`.
%
%   Example usage:
%       [x_den, x_den_smoothed, s, alpha] = common_mode_noise_canceller(x, fs, 0.5, 15, 200, 10:100, true);
%
%   Revision History:
%       2024: First release
%
%   Reza Sameni, 2024
%   The Open-Source Electrophysiological Toolbox
%   https://github.com/alphanumericslab/OSET
%

% Parse optional arguments
if nargin > 2 && ~isempty(varargin{1})
    fc = varargin{1};
else
    fc = [];
end

if nargin > 3 && ~isempty(varargin{2})
    itr = varargin{2};
else
    itr = 10;
end

if nargin > 4 && ~isempty(varargin{3})
    lambda = varargin{3};
else
    lambda = 100;
end

if nargin > 5 && ~isempty(varargin{4})
    optim_indexes = varargin{4};
else
    optim_indexes = 1:size(x, 2);
end

if nargin > 6 && ~isempty(varargin{5})
    plot_results = varargin{5};
else
    plot_results = false;
end

% Baseline subtraction or low-pass filtering, based on `fc`
if ~isempty(fc)
    if fc == 0
        bl = mean(x, 2);
        x = x - bl;
    else
        bl = lp_filter_zero_phase(x, fc/fs);
        x = x - bl;
    end
else
    bl = [];
end

% Initial estimate of common-mode signal `s`
s = median(x, 1);
s = s(:, optim_indexes);

% Iteratively solve for `alpha` and `s`
for k = 1 : itr
    alpha = x(:, optim_indexes) / s;
    s = alpha \ x(:, optim_indexes);
end

% Calculate denoised signal and smoothed version
s = alpha \ x;
x_den = x - alpha * s;

if lambda > 0
    smoothing_order = 2; % 2nd order Tikhonov regularization
    x_den_smoothed = tikhonov_regularization(x_den, smoothing_order, lambda);
else
    x_den_smoothed = x_den;
end

% Optional plotting of results
if plot_results
    figure;
    plot(x');
    hold on;
    plot(s, 'k', 'linewidth', 2);
    grid on;
    xlabel('Time (s)');
    ylabel('Amplitude');
    title('Original Signals with Estimated Common-Mode Signal');
    set(gca, 'fontsize', 14)

    figure;
    subplot(211);
    plot(x');
    grid on;
    title('Original Signals');
    xlabel('Time (s)');
    ylabel('Amplitude');
    set(gca, 'fontsize', 14)

    subplot(212);
    plot(x_den');
    grid on;
    title('Denoised Signals');
    xlabel('Time (s)');
    ylabel('Amplitude');
    set(gca, 'fontsize', 14)
end

File: python/src/oset/generic/common_mode_noise_canceller/common_mode_noise_canceller.py
import numpy as np
import matplotlib.pyplot as plt
from oset.generic.lp_filter import lp_filter_zero_phase, tikhonov_regularization

def common_mode_noise_canceller(x, fs, fc=None, itr=10, lambda_val=100, optim_indexes=None, plot_results=False):
    """
    Estimates and removes common-mode noise from multichannel data.

    Parameters:
        x: Input signal matrix (N channels x T samples).
        fs: Sampling frequency.
        fc: Cutoff frequency for low-pass filtering.
        itr: Number of iterations for the algorithm.
        lambda_val: Regularization parameter for Tikhonov regularization.
        optim_indexes: Indices of the samples to optimize.
        plot_results: Flag to enable/disable result plotting.

    Returns:
        x_den: The denoised signal, with common-mode noise removed.
        x_den_smoothed: The denoised signal after smoothing with Tikhonov regularization.
        s: The estimated common-mode signal.
        alpha: The scaling factor or matrix used in noise cancellation.
        bl: Channel-wise average or baseline estimated by high-pass filtering.
    """
    if optim_indexes is None:
        optim_indexes = np.arange(x.shape[1])

    if fc is not None:
        if fc == 0:
            bl = np.mean(x, axis=1, keepdims=True)
            x = x - bl
        else:
            bl = lp_filter_zero_phase(x, fc/fs)
            x = x - bl
    else:
        bl = np.array([])

    # Initial estimate of common-mode signal `s`
    s = np.median(x, axis=0).reshape(1, -1)
    s = s[:, optim_indexes]

    # print(x.shape)
    # print(bl.shape)
    # print(s.shape)
    

    # Iteratively solve for `alpha` and `s`
    for k in range(itr):
        # print(x[:, optim_indexes].shape)
        # print(s.shape)
        alpha = np.linalg.lstsq(s.T, x[:, optim_indexes].T, rcond=None)[0].T
        s = np.linalg.lstsq(alpha, x[:, optim_indexes], rcond=None)[0]

    # Calculate denoised signal and smoothed version
    s = np.linalg.lstsq(alpha, x, rcond=None)[0]
    x_den = x - np.dot(alpha, s)

    # print(x.shape)
    # print(s.shape)
    # print(alpha.shape)
    # print(x_den.shape)
    
    if lambda_val > 0:
        x_den_smoothed = tikhonov_regularization(x_den, [2], lambda_val)
    else:
        x_den_smoothed = x_den

    # Optional plotting of results
    if plot_results:
        plt.figure()
        plt.plot(x.T)
        plt.hold(True)
        plt.plot(s, 'k', linewidth=2)
        plt.grid(True)
        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')
        plt.title('Original Signals with Estimated Common-Mode Signal')
        plt.gca().set_fontsize(14)

        plt.figure()
        plt.subplot(211)
        plt.plot(x.T)
        plt.grid(True)
        plt.title('Original Signals')
        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')
        plt.gca().set_fontsize(14)

        plt.subplot(212)
        plt.plot(x_den.T)
        plt.grid(True)
        plt.title('Denoised Signals')
        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')
        plt.gca().set_fontsize(14)
        plt.show()

    return x_den, x_den_smoothed, s, alpha, bl

if __name__ == "__main__":
    # Import argparse and set up argument parsing
    import argparse
    parser = argparse.ArgumentParser(description="Estimates and removes common-mode noise from multichannel data.")
    # Add arguments for each parameter, for example:
    parser.add_argument('--fs', type=float, required=True, help='Sampling frequency')
    parser.add_argument('--fc', type=float, default=0.001, help='Cutoff frequency for low-pass filtering')
    parser.add_argument('--itr', type=int, default=10, help='Number of iterations for the algorithm')
    parser.add_argument('--lambda_val', type=float, default=100, help='Regularization parameter for Tikhonov regularization')
    # You'll need to decide how to pass `x` and potentially `optim_indexes`, as they are more complex than simple scalar values

    args = parser.parse_args()

    # Since 'x' and 'optim_indexes' are complex to handle via command-line arguments, you might need to load or define them inside this block
    # For the sake of the example, let's assume 'x' is loaded from a file or defined here directly

    # Example placeholder for loading 'x' and defining 'optim_indexes'
    x = np.random.randn(5, 100)  # Dummy data for demonstration
    optim_indexes = np.arange(25, 75)  # Example index range

    # Call the function with the parsed arguments (and placeholders for 'x' and 'optim_indexes')
    data_den, data_den_smoothed, s, alpha, bl = common_mode_noise_canceller(
        x=x,
        fs=args.fs,
        fc=args.fc,
        itr=args.itr,
        lambda_val=args.lambda_val,
        optim_indexes=optim_indexes,
        plot_results=True  # Assuming you want to plot results when running as a script
    )